### ISSUE TITLE
standard shurtcuts in command mode

### ISSUE THREAD

**Issue by parisni** [neovim/neovim#10799]
- `nvim --version`:
```
NVIM v0.3.8
Build type: Release
LuaJIT 2.1.0-beta3
Compilation: 
Compiled by nixbld
```

Mostly all terminal have those shurcut: 
<ctrl> h : delete char before
<alt> backspace : delete word before
 <alt> b : move cursor a word before
<ctrl> u: delete the entire line
<ctrl> j: enter
<ctrl> p: go up in history
<ctrl n>: go down in history

I d like those shurtcut to be used in the **vim command** mode. Right now,only the last two one can be used,   and only key that works in that command mode are **arrows** and **backspace/delete**.

**Comment by bfredl** [comment]
Also ctrl-h, ctrl-u and ctrl-j already works as you expect. use ctrl-w to delete a word. see `:help c_CTRL-V`  and below for more bindings (including word movements).

**Comment by parisni** [comment]
thanks thats partially coherent with usual terminal shortcut:

ctrl-w for delete a word is ok.
however:

- ctrl-b for going to beginning should be ctrl-a 
- ctrl-right/left or shift-right/left for moving words/char is against vim behavior: the arrows movement are replaced with jklh . It 's not the best to re-introduce them in command isn'it ?

**Comment by bfredl** [comment]
No set of builtin bindings is ever gonna make everyone happy.  You can change them to your taste using `cnoremap <c-a> <c-b>` etc.

**Comment by justinmk** [comment]
I think we should consider promoting https://github.com/tpope/vim-rsi to something "built in", but it will be a lot of work (documentation, tests), and is low-priority. @parisni if you want to take the lead on doing that work then this could move forward.

**Comment by parisni** [comment]
> I think we should consider promoting https://github.com/tpope/vim-rsi to something "built in", 
i didn't know of vim-rsi. Indeed this plugin does exactly what i was describing. To me that's legitimate to add this as built-in.

> @parisni if you want to take the lead on doing that work then this could move forward.
well i m too busy these days, but if nobody does in the future, i would come back and help on this if needed.

**Comment by parisni** [comment]
maybe this one is more relevant with the present issue because it only focuses the command mode:

https://www.vim.org/scripts/script.php?script_id=5633

**Comment by chrisbra** [comment]
what would that need to be built-in and not available as a plugin? After all that's exactly what plugins are for. BTW `ctrl-b` is known vi behaviour (`CTRL-A` is default for the other well-known text editor on Unix).

**Comment by bfredl** [comment]
I suppose it could be a `runtime/pack/dist/opt` shipped plugin. Then it can be opt into with one line of vimrc.

**Comment by parisni** [comment]
> what would that need to be built-in and not available as a plugin? 

plugin or not plugin makes all the difference. An admin system will learn shortcut and expect they work on any vim installation he connects.

learning plugins shurtcuts breaks this interoperability. BTW again, `readline.vim` is far better than `vim-rsi`: 

1. it full complies with readline shorcuts (upper, lower, spaces and so on)
2. it keeps the vim `insert` and `normal` mode intact

it only makes `command` mode far better

**Comment by ryvnf** [comment]
Hi everyone!  I'm Elias, the author of readline.vim.

[readline.vim](https://github.com/ryvnf/readline.vim) is different from similar plugins out there in that it implements a larger subset of Readline commands and each command is carefully implemented to behave exactly like Readline.  I have been using the plugin for years and tested it for many edge cases.  I doubt that there will be many changes from now on.  Latest bug fix was over a year ago and it has been used daily since then.

There are many people who enjoy the plugin.  But I should also note that I've also met many people who think it goes against their Vim workflow, these people generally prefer to use the command-line window (triggered with CTRL-F instead).  So it isn't everyones cup of tea.  I personally can't live without it and think it is a great enhancement of the Vim command-line.  I also encourage anyone interested to try it and read its documentation.  I think many people familiar with Readline would be surprised at how much of an improvement it can be to the command-line in Vim/Neovim.

As a Vim and Neovim user I would be honored to have my plugin included as an optional package.  The plugin is already well documented through a help-file.  And I would be willing to do any work necessary (like writing tests or improve documentation) for it to be included as an optional package.  I would also be happy to answer any question someone might have here about the plugin.

**Comment by justinmk** [comment]
@ryvnf thanks!

> implements a larger subset of Readline commands and each command is carefully implemented to behave exactly like Readline

Perfect readline emulation is secondary to the goal of preserving existing (useful) Vim builtins. (`ctrl-b` is a case where Vim tradition is useless, and readline has precedence). [vim-rsi](https://github.com/tpope/vim-rsi) for example supports readline `CTRL-F` "within a line" and falls back to Vim `c_CTRL-F` at end-of-line.

Another note: mapping to `<esc>x` is not correct in Nvim, and even [this workaround](https://github.com/ryvnf/readline.vim/blob/master/plugin/readline.vim#L114) unnecessarily occupies `<esc>`. Instead, in Nvim, only map `<M-x>` directly (don't use the `<M-.x> => <esc>x` indirection).

**Comment by blueyed** [comment]
> prefer to use the command-line window (triggered with CTRL-F instead)

At least there's a good (fitting) workaround for this (CTRL-X CTRL-E).

It changes at least the behavior of CTRL-B also though (beginning of line in Vim, behaves better with CTRL-A as tmux leader for me).

Maybe default keys overridden could be listed in the doc at least - with the options/settings to disable that?

I've used to use vim-rsi, will use readline.vim now for test driving.

**Comment by ryvnf** [comment]
> Perfect readline emulation is secondary to the goal of preserving existing (useful) Vim builtins. (`ctrl-b` is a case where Vim tradition is useless, and readline has precedence). [vim-rsi](https://github.com/tpope/vim-rsi) for example supports readline `CTRL-F` "within a line" and falls back to Vim `c_CTRL-F` at end-of-line.

I was mostly referring to the way similar commands are implemented. Like `<c-w>` rubbing out a full whitespace delimited word in readline.vim instead of `'iskeyword'` delimited word. Similar for `<m-b>` and `<m-f>` which navigate between space delimited words in rsi.vim, but follow Readlines word boundaries in readline.vim.

The plugin overrides four ordinary Vim shortcuts. These are Ctrl-A, Ctrl-D, Ctrl-F and Ctrl-K. Ctrl-A, Ctrl-D and Ctrl-F all have some equivalent command in Readline (like Ctrl-X Ctrl-E) which is instead used to access the overridden functionality. I think this was a nice way to create an interface consistent with Readline without loosing functionality. How to access each overridden mapping is mentioned in the docs. But I could introduce a list for better overview.

The one command which does not have something equivalent in Readline is Ctrl-K to insert a digraph. In that case the plugin ships with an option to disable the mapping and preserve the default mapping, because there is no logical place to put it. Vim8 also introduced a mapping of Ctrl-T which is only used when incremental search is active. At the moment that mapping is overridden completely with no option to disable it. I could introduce a similar option for that if people want.

> Another note: mapping to `<esc>x` is not correct in Nvim, and even [this workaround](https://github.com/ryvnf/readline.vim/blob/master/plugin/readline.vim#L114) unnecessarily occupies `<esc>`. Instead, in Nvim, only map `<M-x>` directly (don't use the `<M-.x> => <esc>x` indirection).

The `<esc>x` mappings exist purely for compatibility with both Vim and Neovim. Currently in both Vim and Neovim `<esc>` is used in the commands while in Neovim that workaround is used. Then the option `g:readline_esc` controls if Escape should timeout or not. This was to me the cleanest way for me to get the same behavior in both Vim and Nvim without duplicated code to handle each case separately. If desired I can rewrite these portions to entirely follow Neovims convention, but it would require more lines of code and increase complexity in my opinion.

**Comment by ryvnf** [comment]
Also, if this is included as an optional package in Nvim I could simplify the code while still following Nvim's convention for meta mappings by completely removing compatibility with Vim. The disadvantage of that is that one cannot directly copy the package from Nvim into regular Vim anymore.

I think it would be best to let you test-drive the plug-in for a couple of days before any decision is made. I would be happy to clarify things or answer questions in the mean-time. Maybe the issue should also be re-opened as there is an ongoing discussion?

**Comment by justinmk** [comment]
The tradeoffs don't sound good so far.  vim-rsi is [96 lines of code](https://github.com/tpope/vim-rsi/blob/master/plugin/rsi.vim) (even less if legacy workarounds are removed).  That gets us pretty close to readline for 99.999% of users, without spending a lot of time talking about tradeoffs, tests, etc.

**Comment by ryvnf** [comment]
I tried rsi.vim initially as that was what was available.  What immediately threw me off was the fact that I had to press Ctrl-W six times to delete a path like "/path/to/file" when it gets deleted by one press in GNU Readline.  I was also thrown off by the fact that Alt-B and Alt-F move between whitespace delimited words which is also radically different from GNU Readline.

For these reasons rsi.vim didn't help my goal of being able to type command-lines without having to think if I'm in Vim, in bash or some REPL.  So I started to dive into VimL in order to find ways to emulate Readline completely.  This is why the plugin is over 300 lines long.  The code might be a bit hard to follow, but that is also to make VimL correctly handle things like multi-byte and control characters in the command-line, and to make it backwards-compatible with Vim 7.0.

The added complexity was well worth it for me.  I don't think that rsi.vim works for 99.999 % of users interested in this.  readline.vim has 33 stars on GitHub while rsi.vim has 349 which suggest that at-least 10 % of rsi.vim users think readline.vim has something to add.  rsi.vim has also been around for 7 years while readline.vim has existed for 2 years.  rsi.vim also has the added exposure of being a tpope package.  I have great respect for tpope and think he is a great plugin creator.  But I think many people would switch from rsi.vim to readline.vim if they knew it existed.  And I think it would be a bad decision to merge rsi.vim if many current rsi.vim users might find that there is a better alternative down the road.

**Comment by ryvnf** [comment]
I tried rsi.vim initially as that was what was available. What immediately threw me off was the fact that I had to press Ctrl-W six times to delete a path like "/path/to/file" when it gets deleted by one press in GNU Readline. I was also thrown off by the fact that Alt-B and Alt-F move between whitespace delimited words which is also radically different from GNU Readline.

For these reasons rsi.vim didn't help my goal of being able to type command-lines without having to think if I'm in Vim, in bash or some REPL. So I started to dive into VimL in order to find ways to emulate Readline completely. This is why the plugin is over 300 lines long. The code might be a bit hard to follow, but that is also to make VimL correctly handle things like multi-byte and control characters in the command-line, and to make it backwards-compatible with Vim 7.0.

The added complexity was well worth it for me. I don't think that rsi.vim works for 99.999 % of users interested in this. readline.vim has 34 stars on GitHub while rsi.vim has 349 which suggest that at-least 10 % of rsi.vim users think readline.vim has something to add. rsi.vim has also been around for 7 years while readline.vim has existed for 2 years. rsi.vim also has the added exposure of being a tpope package. I have great respect for tpope and think he is a great plugin creator. But I think many people would switch from rsi.vim to readline.vim if they knew it existed. And I think it would be a bad decision to merge rsi.vim if many current rsi.vim users might find that there is a better alternative down the road.
